src/lib/core/func_adapter.h

```diff
@@ -0,0 +1,284 @@
+	 */
+	bool (*is_integer)(struct func_adapter_ctx *ctx);
+	/**
+	 * Pops integer value.
+	 */
+	void (*pop_integer)(struct func_adapter_ctx *ctx, int64_t *integer);
+	/**
+	 * Check if the next returned value is a string.
+	 */
+	bool (*is_string)(struct func_adapter_ctx *ctx);
+	/**
+	 * Pops string value. Argument len is allowed to be NULL.
+	 * Never return NULL.
+	 */
+	void (*pop_string)(struct func_adapter_ctx *ctx, const char **str,
+			   size_t *len);
+	/**
+	 * Check if the next returned value is null or nothing.
+	 */
+	bool (*is_null)(struct func_adapter_ctx *ctx);
+	/**
+	 * Pops null.
+	 */
+	void (*pop_null)(struct func_adapter_ctx *ctx);
+	/**
+	 * Virtual destructor of the class.
+	 */
+	void (*destroy)(struct func_adapter *func);
+};
+
+/**
+ * Base class for all function adapters. Instance of this class should not
+ * be created.
+ * Function is called in several stages:
+ * 1. Preparation - func_adapter_ctx instance is allocated and initialized by
+ *    begin method. Then, all the arguments are pushed in direct order (the
+ *    first argument is pushed first). Pop methods must not be called at this
+ *    stage.
+ * 2. Call - the actual function call. If the call was not successful, it sets
+ *    diag and returns -1. In the case of error, one must stop the calling
+ *    process and call method end to release occupied resources.
+ * 3. Finalization - returned values are popped in direct order (first returned
+ *    value will be popped first). When popping a value of a particular type,
+ *    one must be sure that the next value has this type. It is not necessary to
+ *    pop all the returned values. When all the returned values are popped, all
+ *    the next values will be nulls. After all, method end must be called.
+ */
+struct func_adapter {
+	/**
+	 * Virtual table.
+	 */
+	const struct func_adapter_vtab *vtab;
+};
+
+static inline void
+func_adapter_begin(struct func_adapter *func, struct func_adapter_ctx *ctx)
+{
+	func->vtab->begin(func, ctx);
+}
+
+static inline void
+func_adapter_end(struct func_adapter *func, struct func_adapter_ctx *ctx)
+{
+	func->vtab->end(ctx);
+	TRASH(ctx);
+}
+
+static inline int
+func_adapter_call(struct func_adapter *func, struct func_adapter_ctx *ctx)
+{
+	return func->vtab->call(ctx);
+}
+
+static inline void
+func_adapter_push_number(struct func_adapter *func,
+			 struct func_adapter_ctx *ctx, double number)
+{
+	func->vtab->push_number(ctx, number);
+}
+
+static inline void
+func_adapter_push_integer(struct func_adapter *func,
+			  struct func_adapter_ctx *ctx, int64_t integer)
+{
+	func->vtab->push_integer(ctx, integer);
+}
+
+static inline void
+func_adapter_push_lstring(struct func_adapter *func,
+			  struct func_adapter_ctx *ctx, const char *str,
+			  size_t len)
+{
+	func->vtab->push_string(ctx, str, len);
+}
+
+static inline void
+func_adapter_push_string(struct func_adapter *func,
+			 struct func_adapter_ctx *ctx, const char *str)
```

Comment:
Better call it `push_str0`. Also, there's no point to stick to Lua API naming conventions. Better name the methods `push_str`, `push_str0`, `push_int`, `push_double`. It'd be more consistent with other parts of the Tarantool internals.

Reply:
Done (`push_int` was removed after f2f discussion).